# GENOA cis-meQTL mapping study

 Most existing meQTL mapping studies have focused on individuals of European ancestry and are noticeably underrepresented in other populations, with a particular absence of large studies in populations with African ancestry. Lack of large-scale, well-powered meQTL mapping studies in populations with African ancestry can impede our understanding of the epigenetic mechanisms underlying gene expression and common diseases for these populations. We fill this critical knowledge gap by performing a large-scale in-depth cis-meQTL mapping study using the Illumina EPIC array in 961 African Americans from the Genetic Epidemiology Network of Arteriopathy (GENOA) study.  Our results represent an important step toward revealing the causal/mediation role of methylation underlying gene expression, facilitating the functional integration and interpretation of epigenetic and gene regulatory changes underlying common diseases in African Americans.
 
# Data Availability

The cis-meQTL mapping summary statistics could be found at http://xzlab.org/data.html

The SNP data is available from the Database of Genotypes and Phenotypes (dbGaP Study Accession: phs001238.v2.p1). 

Data can also be obtained from the corresponding author on reasonable request.

1.meQTL_mapping.md
2.conditional_analysis.md
3.meQTL_results_comparison.md
4.colocalization.md
5.mediation_analysis.md
6.Figures_for_meQTL.md

# Analysis codes
| File name  | Usage |
| ------------- | ------------- |
|1.meQTL_mapping.md | for each CpG, we map SNPs within +/-50kb of the CpG site and extract the plink file for genotype  |
|2.conditional_analysis.md | conditional analysis, efficient codes and scripts |
|3.meQTL_results_comparison.md|Compare results with previous studies|
|4.colocalization.md | Codes for colocalization and molocalization|
|5.mediation_analysis.md | Codes for mediation analysis|
|6.Figures_for_meQTL.md| codes for figures in paper|
